lib/obj.js

```diff
@@ -1,365 +1,447 @@
-                    result[10], result[11], undefined, result[12]
-                );
-            } else if (/^usemtl /.test(line)) {
-                var materialName = line.substring(7).trim();
-                useMaterial(materialName);
-            }
+    // Create a default node in case there are no o/g/usemtl lines in the obj
+    addNode();
+
+    // Parse the obj file
+    return readLines(objPath, parseLine)
+        .then(function() {
+            // Unload resources
+            positions = undefined;
+            normals = undefined;
+            uvs = undefined;
+
+            // Load materials and images
+            return finishLoading(nodes, mtlPaths, objPath);
         });
+}
 
-        stream.on('end', function () {
-            resolve({
-                vertexCount: vertexCount,
-                vertexArray: vertexArray,
-                positionMin: positionMin,
-                positionMax: positionMax,
-                hasUVs: hasUVs,
-                hasNormals: hasNormals,
-                materialGroups: materialGroups,
-                materials: materials,
-                images: images
-            });
+function finishLoading(nodes, mtlPaths, objPath) {
+    nodes = cleanNodes(nodes);
+    if (nodes.length === 0) {
+        throw new RuntimeError(objPath + ' does not have any geometry data');
+    }
+    return loadMaterials(mtlPaths, objPath)
+        .then(function(materials) {
+            var imagePaths = getImagePaths(materials);
+            return loadImages(imagePaths, objPath)
+                .then(function(images) {
+                    return {
+                        nodes : nodes,
+                        materials : materials,
+                        images : images
+                    };
+                });
         });
+}
+
+function getAbsolutePath(mtlPath, objPath) {
+    if (!path.isAbsolute(mtlPath)) {
+        mtlPath = path.join(path.dirname(objPath), mtlPath);
+    }
+    return mtlPath;
+}
+
+function loadMaterials(mtlPaths, objPath) {
+    var materials = {};
+    return Promise.map(mtlPaths, function(mtlPath) {
+        mtlPath = getAbsolutePath(mtlPath, objPath);
+        return loadMtl(mtlPath)
+            .then(function(materialsInMtl) {
+                materials = combine(materials, materialsInMtl);
+            });
+    }).then(function() {
+        return materials;
+    });
+}
+
+function loadImages(imagePaths) {
+    var images = {};
+    return Promise.map(imagePaths, function(imagePath) {
+        return loadImage(imagePath)
+            .then(function(image) {
+                if (defined(image)) {
+                    images[imagePath] = image;
+                }
+            });
+    }).then(function() {
+        return images;
     });
 }
 
-function getImages(inputPath, materials) {
-    // Collect all the image files from the materials
-    var images = [];
+function getImagePaths(materials) {
+    var imagePaths = [];
```

Comment:
Does `indexOf` do a linear search? Maybe a dictionary be better here for an `.obj` with a huge number of textures.